File: python/3871.count-commas-in-range-ii.py
class Solution:
    def countCommas(self, n: int) -> int:
        res = 0
        e = 1000  # exponent 1e3, 1e6, 1e9, 1e12 ...
        p = 1  #  p (power), count of commas for exponent, 1,2,3,4...

        while e <= n:
            x = min(n + 1, e * 1000) # next exponent or the limit n (+1 inclusive)
            res += (x - e) * p

            e *= 1000
            p += 1

        return res

File: cpp/3871.count-commas-in-range-ii.cpp
class Solution {
public:
    long long countCommas(long long n) {
        long long res = 0;
        long long exp = 1000; // exponent: 1e3, 1e6, 1e9, 1e12, ...
        long long nexp = exp * 1000; // next exponent
        long long p = 1; // power: 1, 2, 3, 4, ...

        while (exp <= n) {
          nexp = std::min(n + 1L, exp * 1000L); // nexp is the next exponent (or n inclusive)
          
          res += (nexp - exp) * p;

          exp = nexp;
          p += 1;
        }

        return res;
    }
};